diff --git a/include/vulkan/vk_android_native_buffer.h b/include/vulkan/vk_android_native_buffer.h
index 80e5d26c8ee..cf17acd99db 100644
--- a/include/vulkan/vk_android_native_buffer.h
+++ b/include/vulkan/vk_android_native_buffer.h
@@ -1,5 +1,5 @@
 /* MESA: A hack to avoid #ifdefs in driver code. */
-#ifdef __ANDROID__
+#if 0 //defined(__ANDROID__) || !defined(__MICEWINE__)
 
 #include <cutils/native_handle.h>
 #if ANDROID_API_LEVEL < 28
diff --git a/src/util/detect_os.h b/src/util/detect_os.h
index 86286dfbe70..49195b3398c 100644
--- a/src/util/detect_os.h
+++ b/src/util/detect_os.h
@@ -25,7 +25,7 @@
  * also be defined.
  */
 #if defined(__ANDROID__)
-#define DETECT_OS_ANDROID 1
+#define DETECT_OS_ANDROID 0
 #endif
 
 #if defined(__FreeBSD__) || defined(__FreeBSD_kernel__)
