diff -uNr 49-libevent/evdns.c 49-libevent.mod/evdns.c
--- 49-libevent/evdns.c	2020-07-05 09:02:46.000000000 -0300
+++ 49-libevent.mod/evdns.c	2024-03-25 20:14:08.375999228 -0300
@@ -52,6 +52,9 @@
 #include "evconfig-private.h"
 
 #include <sys/types.h>
+#ifdef __ANDROID__
+#include <sys/system_properties.h>
+#endif
 
 #ifndef _FORTIFY_SOURCE
 #define _FORTIFY_SOURCE 3
@@ -3655,7 +3658,7 @@
 	evutil_snprintf(path_out, len_out, "%s%s", path, hostfile);
 	return path_out;
 #else
-	return mm_strdup("/etc/hosts");
+	return mm_strdup("/data/data/com.micewine.emu/files/usr/etc/hosts");
 #endif
 }
 
@@ -4032,6 +4035,21 @@
 
 #ifdef _WIN32
 		r = evdns_base_config_windows_nameservers(base);
+#elif defined(__ANDROID__)
+		/** From: http://www.programering.com/a/MjMwcjMwATA.html */
+		int add_servers = 0;
+		char buf[PROP_VALUE_MAX];
+		r = __system_property_get("net.dns1", buf);
+		if (r >= 7) {
+			add_servers++;
+			evdns_base_nameserver_ip_add(base, buf);
+		}
+		r = __system_property_get("net.dns2", buf);
+		if (r >= 7) {
+			add_servers++;
+			evdns_base_nameserver_ip_add(base, buf);
+		}
+		if (add_servers == 0) evdns_base_nameserver_ip_add(base, "8.8.8.8");
 #else
 		r = evdns_base_resolv_conf_parse(base, opts, "/etc/resolv.conf");
 #endif
diff -uNr 49-libevent/evutil.c 49-libevent.mod/evutil.c
--- 49-libevent/evutil.c	2020-07-05 09:02:46.000000000 -0300
+++ 49-libevent.mod/evutil.c	2024-03-25 20:08:52.615984829 -0300
@@ -49,6 +49,7 @@
 #include <sys/socket.h>
 #endif
 #ifdef EVENT__HAVE_UNISTD_H
+#define __USE_GNU 1 /* For pipe2(). */
 #include <unistd.h>
 #endif
 #ifdef EVENT__HAVE_FCNTL_H
