diff --git a/bin/install_megadrivers.py b/bin/install_megadrivers.py
index 60ee4576c60..1e7108035cd 100644
--- a/bin/install_megadrivers.py
+++ b/bin/install_megadrivers.py
@@ -24,7 +24,16 @@
 
 import argparse
 import os
+import shutil
 
+def link(src, dest):
+    shutil.copyfile(src, dest)
+ 
+def unlink(src):
+    os.remove(src)
+
+os.link = link
+os.unlink = unlink
 
 def main():
     parser = argparse.ArgumentParser()
