diff --git a/src/vulkan/wrapper/wrapper_instance.c b/src/vulkan/wrapper/wrapper_instance.c
index 8a2aba85b33..dcfda2da086 100644
--- a/src/vulkan/wrapper/wrapper_instance.c
+++ b/src/vulkan/wrapper/wrapper_instance.c
@@ -5,14 +5,6 @@
 #include "vk_dispatch_table.h"
 #include "vk_extensions.h"
 
-uint64_t WRAPPER_DEBUG;
-
-static const struct debug_control debug_control[] = {
-   { "placed",       WRAPPER_MAP_MEMORY_PLACED },
-   { "bc",           WRAPPER_BC },
-   { NULL, },
-};
-
 const struct vk_instance_extension_table wrapper_instance_extensions = {
    .KHR_get_surface_capabilities2 = true,
    .EXT_surface_maintenance1 = true,
@@ -61,9 +53,6 @@ static bool vulkan_library_init()
    if (vulkan_library_handle)
       return true;
 
-   WRAPPER_DEBUG = parse_debug_string(getenv("WRAPPER_DEBUG"),
-                                      debug_control);
-
    const char *env = getenv("WRAPPER_VULKAN_PATH");
    vulkan_library_handle = dlopen(env ? env : DEFAULT_VULKAN_PATH,
                                   RTLD_LOCAL | RTLD_NOW);
diff --git a/src/vulkan/wrapper/wrapper_physical_device.c b/src/vulkan/wrapper/wrapper_physical_device.c
index a2a8ac387d1..3d109bf4d3c 100644
--- a/src/vulkan/wrapper/wrapper_physical_device.c
+++ b/src/vulkan/wrapper/wrapper_physical_device.c
@@ -127,16 +127,12 @@ VkResult enumerate_physical_device(struct vk_instance *_instance)
       supported_features->swapchainMaintenance1 = true;
       supported_features->imageCompressionControlSwapchain = false;
 
-      pdevice->enable_bc =
-         !supported_features->textureCompressionBC
-         && (WRAPPER_DEBUG & WRAPPER_BC);
+      pdevice->enable_bc = true;
 
       if (pdevice->enable_bc)
          supported_features->textureCompressionBC = true;
 
-      pdevice->enable_map_memory_placed =
-         !pdevice->vk.supported_extensions.EXT_map_memory_placed
-         && (WRAPPER_DEBUG & WRAPPER_MAP_MEMORY_PLACED);
+      pdevice->enable_map_memory_placed = true;
 
       if (pdevice->enable_map_memory_placed) {
          pdevice->vk.supported_extensions.EXT_map_memory_placed = true;
diff --git a/src/vulkan/wrapper/wrapper_private.h b/src/vulkan/wrapper/wrapper_private.h
index a7a9c22e627..d45746cf6bf 100644
--- a/src/vulkan/wrapper/wrapper_private.h
+++ b/src/vulkan/wrapper/wrapper_private.h
@@ -12,11 +12,6 @@ extern const struct vk_instance_extension_table wrapper_instance_extensions;
 extern const struct vk_device_extension_table wrapper_device_extensions;
 extern const struct vk_device_extension_table wrapper_filter_extensions;
 
-extern uint64_t WRAPPER_DEBUG;
-
-#define WRAPPER_MAP_MEMORY_PLACED      (1ull << 0)
-#define WRAPPER_BC                     (1ull << 1)
-
 struct wrapper_instance {
    struct vk_instance vk;
 
