diff --git a/src/overlay.cpp b/src/overlay.cpp
index a1cfd25..fa326d4 100644
--- a/src/overlay.cpp
+++ b/src/overlay.cpp
@@ -765,7 +765,11 @@ void init_system_info(){
       trim(cpu);
       kernel = exec("uname -r");
       trim(kernel);
+   #ifdef __ANDROID__
+      os = "Android";
+   #else
       os = exec("sed -n 's/PRETTY_NAME=\\(.*\\)/\\1/p' /etc/*-release");
+   #endif
       os.erase(remove(os.begin(), os.end(), '\"' ), os.end());
       trim(os);
       cpusched = read_line("/sys/devices/system/cpu/cpu0/cpufreq/scaling_governor");
