--- 62-glib/glib/gutils.c	2024-05-08 12:29:04.000000000 -0300
+++ 62-glib.mod/glib/gutils.c	2024-05-11 19:28:55.815718168 -0300
@@ -385,7 +385,7 @@
        * what to search if PATH is unset. POSIX may, dunno.
        */
       
-      path = "/bin:/usr/bin:.";
+      path = "/data/data/com.micewine.emu/files/usr/bin:.";
     }
 #else
   if G_LIKELY (original_path == NULL)
@@ -1018,7 +1018,7 @@
       if (tmp == NULL || *tmp == '\0')
         {
           g_free (tmp);
-          tmp = g_strdup ("/tmp");
+          tmp = g_strdup ("/data/data/com.micewine.emu/files/usr/tmp");
         }
 #endif /* !G_OS_WIN32 */
 
@@ -1680,7 +1680,7 @@
   else
     return NULL;
 #elif defined (G_OS_UNIX)
-  const gchar * const os_release_files[] = { "/etc/os-release", "/usr/lib/os-release" };
+  const gchar * const os_release_files[] = { "/data/data/com.micewine.emu/files/usr/etc/os-release", "/data/data/com.micewine.emu/files/usr/lib/os-release" };
   gsize i;
   gchar *buffer = NULL;
   gchar *result = NULL;
@@ -2695,7 +2695,7 @@
    */
 #ifndef G_OS_WIN32
   if (!data_dirs || !data_dirs[0])
-    data_dirs = "/usr/local/share/:/usr/share/";
+    data_dirs = "/data/data/com.micewine.emu/files/usr/local/share/:/data/data/com.micewine.emu/files/usr/share/";
 
   data_dir_vector = g_strsplit (data_dirs, G_SEARCHPATH_SEPARATOR_S, 0);
 #else
