diff -uNr ../mangohud/src/gpu.cpp ./src/gpu.cpp
--- ../mangohud/src/gpu.cpp	2025-03-05 06:56:10.000000000 -0300
+++ ./src/gpu.cpp	2025-06-26 19:20:46.145938572 -0300
@@ -20,6 +20,10 @@
 namespace fs = ghc::filesystem;
 
 GPUS::GPUS(overlay_params* params) : params(params) {
+#ifdef __ANDROID__
+    return;
+#endif
+
     std::vector<std::string> gpu_entries;
 
     for (const auto& entry : fs::directory_iterator("/sys/class/drm")) {
