diff --git a/include/vulkan/vk_android_native_buffer.h b/include/vulkan/vk_android_native_buffer.h
index 21f82ef57a2..5c67746ca44 100644
--- a/include/vulkan/vk_android_native_buffer.h
+++ b/include/vulkan/vk_android_native_buffer.h
@@ -17,7 +17,7 @@
  * system rather the build target.
  */
 
-#if defined(__ANDROID__) || defined(ANDROID)
+#if 0
 
 #include <cutils/native_handle.h>
 #if ANDROID_API_LEVEL < 28
diff --git a/src/util/detect_os.h b/src/util/detect_os.h
index 86286dfbe70..49195b3398c 100644
--- a/src/util/detect_os.h
+++ b/src/util/detect_os.h
@@ -25,7 +25,7 @@
  * also be defined.
  */
 #if defined(__ANDROID__)
-#define DETECT_OS_ANDROID 1
+#define DETECT_OS_ANDROID 0
 #endif
 
 #if defined(__FreeBSD__) || defined(__FreeBSD_kernel__)
