diff --git a/src/vulkan/wrapper/wrapper_device_memory.c b/src/vulkan/wrapper/wrapper_device_memory.c
index f91dcf2a54a..1f1846c025f 100644
--- a/src/vulkan/wrapper/wrapper_device_memory.c
+++ b/src/vulkan/wrapper/wrapper_device_memory.c
@@ -1,6 +1,10 @@
+#define native_handle_t __native_handle_t
+#define buffer_handle_t __buffer_handle_t
 #include "wrapper_private.h"
 #include "wrapper_entrypoints.h"
 #include "vk_common_entrypoints.h"
+#undef native_handle_t
+#undef buffer_handle_t
 #include "util/os_file.h"
 #include "vk_util.h"

