diff --git a/configure.ac b/configure.ac
index 1c59651ba1e..98298efd73f 100644
--- a/configure.ac
+++ b/configure.ac
@@ -710,23 +710,10 @@ case $host_os in
     DLLFLAGS="$DLLFLAGS -fPIC"
     LDDLLFLAGS="-fPIC"
     LDEXECFLAGS="-Wl,-pie"
-    enable_wineandroid_drv=${enable_wineandroid_drv:-yes}
+    enable_winex11_drv=${enable_winex11_drv:-yes}
     WINE_TRY_CFLAGS([-Wl,-z,defs],[UNIXLDFLAGS="$UNIXLDFLAGS -Wl,-z,defs"])
     WINE_TRY_CFLAGS([-fPIC -Wl,--export-dynamic],[WINELOADER_LDFLAGS="-Wl,--export-dynamic"])
     WINEPRELOADER_LDFLAGS="-static -nostartfiles -nodefaultlibs -Wl,-Ttext=0x7d400000"
-
-    WINE_CHECK_SONAME(EGL,eglGetProcAddress)
-    WINE_CHECK_SONAME(GLESv2,glFlush)
-
-    if test "x$exec_prefix" = xNONE
-    then
-      case $HOST_ARCH in
-        i386)    exec_prefix='${prefix}/x86' ;;
-        x86_64)  exec_prefix='${prefix}/x86_64' ;;
-        arm)     exec_prefix='${prefix}/armeabi-v7a' ;;
-        aarch64) exec_prefix='${prefix}/arm64-v8a' ;;
-      esac
-    fi
     ;;
 
   *)
