diff -uNr 39-Vulkan-Tools/cube/CMakeLists.txt 39-Vulkan-Tools.mod/cube/CMakeLists.txt
--- 39-Vulkan-Tools/cube/CMakeLists.txt	2024-03-08 14:39:14.000000000 -0300
+++ 39-Vulkan-Tools.mod/cube/CMakeLists.txt	2024-03-17 12:59:36.145975173 -0300
@@ -130,8 +130,6 @@
 
 if(WIN32)
     add_definitions(-DVK_USE_PLATFORM_WIN32_KHR -DWIN32_LEAN_AND_MEAN -DNOMINMAX)
-elseif(ANDROID)
-    add_definitions(-DVK_USE_PLATFORM_ANDROID_KHR)
 elseif(APPLE)
     add_definitions(-DVK_USE_PLATFORM_METAL_EXT)
 elseif(CMAKE_SYSTEM_NAME MATCHES "Linux|BSD")
diff -uNr 39-Vulkan-Tools/cube/cube.c 39-Vulkan-Tools.mod/cube/cube.c
--- 39-Vulkan-Tools/cube/cube.c	2024-03-08 14:39:14.000000000 -0300
+++ 39-Vulkan-Tools.mod/cube/cube.c	2024-03-17 12:59:36.149308505 -0300
@@ -97,7 +97,7 @@
     fflush(stdout);
 }
 
-#elif defined __ANDROID__
+#elif !defined(__ANDROID__)
 #include <android/log.h>
 #define ERR_EXIT(err_msg, err_class)                                           \
     do {                                                                       \
@@ -534,7 +534,7 @@
     if (!demo->suppress_popups) MessageBox(NULL, message, "Alert", MB_OK);
     in_callback = false;
 
-#elif defined(ANDROID)
+#elif !defined(__ANDROID__)
 
     if (messageSeverity & VK_DEBUG_UTILS_MESSAGE_SEVERITY_INFO_BIT_EXT) {
         __android_log_print(ANDROID_LOG_INFO, APP_SHORT_NAME, "%s", message);
@@ -4257,7 +4257,7 @@
             continue;
         }
 
-#if defined(ANDROID)
+#if !defined(ANDROID)
         ERR_EXIT("Usage: vkcube [--validate]\n", "Usage");
 #else
         char *message =
